language/bytecode-verifier/src/signature.rs

```diff
@@ -7,284 +7,212 @@
-            .chain(self.verify_function_signatures())
-            .chain(self.verify_fields())
-            .chain(self.verify_code_units())
-            .collect()
+    pub fn verify(mut self) -> VMResult<()> {
+        self.verify_signature_pool()?;
+        self.verify_function_signatures()?;
+        self.verify_fields()?;
+        self.verify_code_units()
     }
 
-    fn verify_signatures(&self) -> impl Iterator<Item = VMStatus> + '_ {
-        use SignatureToken::*;
-
-        self.module.signatures().iter().filter_map(|sig| {
-            for ty in &sig.0 {
-                match ty {
-                    Reference(inner) | MutableReference(inner) => match **inner {
-                        Reference(_) | MutableReference(_) => {
-                            return Some(VMStatus::new(StatusCode::INVALID_SIGNATURE_TOKEN));
-                        }
-                        _ => (),
-                    },
-                    Vector(inner) => match **inner {
-                        Reference(_) | MutableReference(_) => {
-                            return Some(VMStatus::new(StatusCode::INVALID_SIGNATURE_TOKEN));
-                        }
-                        _ => (), // TODO: vector of vectors
-                    },
-                    _ => (),
-                }
-            }
-            None
-        })
+    // Review: Is this really necessary? Do we care about checking unused signatures?
+    fn verify_signature_pool(&mut self) -> VMResult<()> {
+        self.signature_context = None;
+        for i in 0..self.module.signatures().len() {
+            self.check_signature(SignatureIndex::new(i as TableIndex))?
+        }
+        Ok(())
     }
 
-    fn verify_function_signatures(&self) -> impl Iterator<Item = VMStatus> + '_ {
-        self.module
-            .function_handles()
-            .iter()
-            .enumerate()
-            .flat_map(move |(idx, fh)| {
-                let type_parameters = fh.type_parameters.as_slice();
-
-                let return_ = self.module.signature_at(fh.return_);
-                let errors_return = return_.0.iter().flat_map(move |ty| {
-                    self.check_signature(ty, type_parameters)
-                        .into_iter()
-                        .map(move |err| append_err_info(err, IndexKind::Signature, idx))
-                });
-                let parameters = self.module.signature_at(fh.parameters);
-                let errors_params = parameters.0.iter().flat_map(move |ty| {
-                    self.check_signature(ty, type_parameters)
-                        .into_iter()
-                        .map(move |err| append_err_info(err, IndexKind::Signature, idx))
-                });
-                errors_return.chain(errors_params)
-            })
+    fn verify_function_signatures(&mut self) -> VMResult<()> {
+        for (idx, fh) in self.module.function_handles().iter().enumerate() {
+            self.signature_context = Some(fh.type_parameters.clone());
```

Comment:
I need some help to understand what this is doing. I don't see the usage of this so asking...